File: question/neetcode150/sliding_window/permutation-in-string.py
from collections import defaultdict

class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        l = 0
        d = defaultdict(int)

        for c in s1:
            d[c]+=1
        
        for r in range(len(s2)):
            d[s2[r]]-=1
            if d[s2[r]] == 0:
                del d[s2[r]]
            if r-l>=len(s1):
                d[s2[l]]+=1
                if d[s2[l]] == 0:
                    del d[s2[l]]
                l+=1

            if not d:
                return True
                
        return False

File: question/neetcode150/sliding_window/permutation-in-string.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        unordered_map<char, int> d;
        for (auto c: s1){
            d[c]++;
        }

        int n = s2.size();
        int n1 = s1.size();

        for (int r=0;r<n;r++){
            char c = s2[r];
            d[c]--;
            if(d[c]==0){
                d.erase(c);
            }

            if(r>=n1){
                char c = s2[r-n1];
                d[c]++;
                if (d[c]==0)
                    d.erase(c);
            }

            if (!d.size()){
                return true;
            }
        }

        return false;
    }
};